src/app/features/reporting/reporting.service.ts

```diff
@@ -0,0 +1,172 @@
+import { Injectable } from "@angular/core";
+import { QueryService } from "./query.service";
+
+export interface Aggregation {
+  query: string;
+  groupBy?: string[];
+  label?: string;
+  aggregations?: Aggregation[];
+}
+
+export interface ReportRow {
+  header: { label: string; values: any[]; result: number };
+  subRows: ReportRow[];
+}
+
+@Injectable({
+  providedIn: "root",
+})
+export class ReportingService {
+  private aggregations: Aggregation[] = [];
+  private fromDate: Date;
+  private toDate: Date;
+
+  constructor(private queryService: QueryService) {}
+
+  public setAggregations(aggregations: Aggregation[]) {
+    this.aggregations = aggregations;
+  }
+
+  public calculateReport(from?: Date, to?: Date): Promise<ReportRow[]> {
+    this.fromDate = from;
+    this.toDate = to;
+    return this.calculateAggregations(this.aggregations);
+  }
+
+  private async calculateAggregations(
+    aggregations: Aggregation[] = [],
+    data?: any[],
+    additionalValues: any[] = []
+  ): Promise<ReportRow[]> {
+    const resultRows: ReportRow[] = [];
+    let currentSubRows = resultRows;
+    for (const aggregation of aggregations) {
+      const queryResult = await this.queryService.queryData(
+        aggregation.query,
+        this.fromDate,
+        this.toDate,
+        data
+      );
+      if (aggregation.label) {
+        const newRow = {
+          header: {
+            label: aggregation.label,
+            values: additionalValues,
```

Comment:
does it make sense to call these "values"? They seem rather be grouping/filter criteria?